### PR TITLE
APIS08c - Add and edit authorized application

### LINKED ISSUE #30525: APIS08c - Authorized Application - Edit an Authorized Application
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

This is a new feature. It aims to make the Authorized Apps manageable in the BO

### Describe the solution you'd like

As a shop administrator
I want to be able to edit an authorized application in the Authz server management page in the BO
In order to change my authorized app parameters

### Alternatives you've considered

_No response_

### Additional context

[Figma link](https://cutt.ly/2MUXEgy)

Acceptance test
AT1

1. Given that
   1. I am a shop administrator
   2. The authorization server administration is available
2. When I open the the authz server administration page in the BO
3. Then
   1. I can see the list of applications that have an API access to my shop
   2. I can click on "Edit" to edit an authorized app

### Do you plan to work on this feature?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by mflasquin**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add and edit authorized application
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Try to add / edit authorized application
| Fixed ticket?     | Fixes #30525
| Related PRs       | -

### CODE REVIEW

File: src/Core/Domain/AuthorizationServer/CommandHandler/EditApplicationHandler.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @throws DuplicateApplicationNameException|ApplicationNotFoundException
+     */
+    public function handle(EditApplicationCommand $command): void
+    {
+        /** @var AuthorizedApplication $application */
+        $application = $this->applicationRepository->getById($command->getApplicationId());
+
+        if ($application === null) {
+            throw new ApplicationNotFoundException(sprintf('Application with id "%d" was not found.', $command->getApplicationId()->getValue()));
```
Comment (FabienPapet) [inline src/Core/Domain/AuthorizationServer/CommandHandler/EditApplicationHandler.php]: You should throw this exception in the repository. Null checks can then be removed (see your PHPStan error)

File: src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml [hunk 2]
```diff
@@ -2167,3 +2167,9 @@ services:
     public: true
     tags:
       - { name: form.type }
+
+  PrestaShopBundle\Form\Admin\AdvancedParameters\AuthorizationServer\ApplicationType:
+    parent: 'form.type.translatable.aware'
+    public: true
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:2196]:
```suggestion
    public: false
```

File: src/PrestaShopBundle/Resources/config/services/core/domain/authorization_server.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+services:
+  _defaults:
+    public: false
+    autowire: true
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/core/domain/authorization_server.yml:4]: ❤️

**Review by FabienPapet — CHANGES_REQUESTED** [review]

**Comment by mflasquin** [comment]
Hi @jolelievre, @eternoendless, @atomiix can you take a look on this PR please :smile:

File: src/Core/Domain/AuthorizationServer/CommandHandler/AddApplicationHandlerInterface.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * Defines contract for AddApplicationHandler
+ */
+interface AddApplicationHandlerInterface
+{
+    /**
+     * @param AddApplicationCommand $command
+     *
+     * @return ApplicationIdInterface
+     */
+    public function handle(AddApplicationCommand $command): ApplicationIdInterface;
```
Comment (atomiix) [inline src/Core/Domain/AuthorizationServer/CommandHandler/AddApplicationHandlerInterface.php]: Why using a interface for the ValueObject?

File: src/Core/Domain/AuthorizationServer/ValueObject/ApplicationIdInterface.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\AuthorizationServer\ValueObject;
+
+/**
+ * @see ApplicationId
+ */
+interface ApplicationIdInterface
```
Comment (eternoendless) [inline src/Core/Domain/AuthorizationServer/ValueObject/ApplicationIdInterface.php]: IMO it makes no sense to add an interface to a ValueObject. A value object is supposed to work like a data type. Would you use StringInterface or ArrayInterface? No, because nothing can replace them.

File: src/Core/Domain/AuthorizationServer/Command/EditApplicationCommand.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        return $this->name;
+    }
+
+    /**
+     * @param string $name
+     *
+     * @return self
+     */
+    public function setName(string $name): self
```
Comment (eternoendless) [inline src/Core/Domain/AuthorizationServer/Command/EditApplicationCommand.php:86]: In general, commands should be immutable. See https://cqrs.nu/Faq/commands-and-events#:~:text=Commands%20are%20immutable%20because%20their,took%20place%20in%20the%20past.

File: src/Core/Domain/AuthorizationServer/Exception/DuplicateApplicationNameException.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Is thrown when Application constraint is violated
+ */
+class DuplicateApplicationNameException extends ApplicationException
+{
+    /** @var string */
+    private $name;
+
+    public function __construct(string $name, $message = '', $code = 0, \Throwable $previous = null)
```
Comment (eternoendless) [inline src/Core/Domain/AuthorizationServer/Exception/DuplicateApplicationNameException.php]: What is name supposed to be? --> add a comment or rename to `$duplicateApplicationName`

File: src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/AuthorizationServer/Application/Blocks/form.html.twig [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+{{ form_start(applicationForm) }}
+<div class="card">
+  <h3 class="card-header">
+    {{ 'Application information'|trans({}, 'Admin.Catalog.Feature') }}
+  </h3>
+  <div class="card-body">
+    <div class="form-wrapper">
+      {{ form_row(applicationForm.name) }}
+      {{ form_row(applicationForm.description) }}
+      {{ form_widget(applicationForm) }}
```
Comment (eternoendless) [inline src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/AuthorizationServer/Application/Blocks/form.html.twig]: `{{ form_widget(applicationForm) }}` Should be the only call here

See: https://github.com/PrestaShop/PrestaShop/issues/16482

File: src/Core/Domain/AuthorizationServer/CommandHandler/AddApplicationHandlerInterface.php (hunk 4 above)
Comment (mflasquin) [reply src/Core/Domain/AuthorizationServer/CommandHandler/AddApplicationHandlerInterface.php]: I think I've take the wrong example (ManufacturerIdInterface) so in this case there is no reason, you're right

File: src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/AuthorizationServer/Application/create.html.twig [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+{% extends '@PrestaShop/Admin/layout.html.twig' %}
+
+{% block content %}
+  {% include '@PrestaShop/Admin/Configure/AdvancedParameters/AuthorizationServer/Application/Blocks/form.html.twig' %}
+{% endblock %}
+
+{% block javascripts %}
+  {{ parent() }}
+  <script src="{{ asset('themes/new-theme/public/authorization-server.bundle.js') }}"></script>
+{% endblock %}
```
Comment (atomiix) [inline src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/AuthorizationServer/Application/create.html.twig]: Maybe you can remove this? Unless you know there will be a JS file coming in another PR?

File: src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/AuthorizationServer/Application/edit.html.twig [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+{% extends '@PrestaShop/Admin/layout.html.twig' %}
+
+{% block content %}
+  {% include '@PrestaShop/Admin/Configure/AdvancedParameters/AuthorizationServer/Application/Blocks/form.html.twig' %}
+{% endblock %}
+
+{% block javascripts %}
+  {{ parent() }}
+  <script src="{{ asset('themes/new-theme/public/authorization-server.bundle.js') }}"></script>
+{% endblock %}
```
Comment (atomiix) [inline src/PrestaShopBundle/Resources/views/Admin/Configure/AdvancedParameters/AuthorizationServer/Application/edit.html.twig]: Same here?

**Review by atomiix — DISMISSED** [review]

**Review by FabienPapet — DISMISSED** [review]

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @mflasquin 

I have checked your PR and I found this issue : 

An exception is thrown when adding a new authorized application without putting a description 

image


https://user-images.githubusercontent.com/92912932/214062242-726dc625-b86c-4ad3-810d-75c662f08a1a.mp4


Could you please check it?

Thank You!

**Review by FabienPapet — DISMISSED** [review]

**Review by nicosomb — DISMISSED** [review]

**Review by FabienPapet — DISMISSED** [review]

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @mflasquin 

I have checked again your PR. I found these issues : 

**Issue 1 :** The Application name should be : `App name ` and not `Name` as described in the specifications 

**Figma :** 
image


**With PR :** 
image


**Issue 2 :** The `Help` button should be in the right position when the application name contains 255 characters 

image


**Issue 3 : Question:** The application name can contain specific characters. Is that the expected behavior ?

image

ping @fabienvallon What do you think?

**Issue 4:** The application name created with a new shop is displayed also with the default shop. Is that the expected behavior?

-Enable multistore
-Add a new shop
-Go to Authorization Server > Select the new shop already created 
-Add a new application with this new shop
-Change to the default shop 
-See error 

image

ping @fabienvallon What do you think?


Could you please check it ? Waiting for your feedback!

Thank you!

**Comment by fabienvallon** [comment]
Yes it can contain specific characters for the AppName.
We should just need to be sure it is not a security issue ( XSS....)

**Review by atomiix — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Comment by fabienvallon** [comment]
About multishop : 
Short term ( read now ) => do not test / enable 

mid-term ( post 8.1) => we will implement it yes 

So it should not be a blocker right now

**Review by FabienPapet — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Review by HanaRebaiQA — APPROVED** [review]
Hello @mflasquin 

I have checked again this PR. The mentioned issues are well fixed. Thanks for that!
-Check the automatic tests: OK
-Check the add of an Application without a description: OK
-Check the add of an Application with a description: OK
-Check the add of an Application with a long name > 255 characters: OK
-Check the edit of an application: OK
-Check the duplicate of an application: OK
-Check the pagination : OK

So, It's QA :heavy_check_mark: 

Thank You!

**Comment by atomiix** [comment]
Thanks @mflasquin and @HanaRebaiQA!
